services/catalog/src/simcore_service_catalog/api/routes/services_resources.py

```diff
@@ -25,104 +33,175 @@
 router = APIRouter()
 logger = logging.getLogger(__name__)
 
-
-def _parse_generic_resource(
-    generic_resources: list[Any], service_resources: ServiceResources
-) -> None:
-    for res in generic_resources:
-        if not isinstance(res, dict):
-            continue
-
-        if named_resource_spec := res.get("NamedResourceSpec"):
-            service_resources.setdefault(
-                named_resource_spec["Kind"],
-                ResourceValue(limit=0, reservation=named_resource_spec["Value"]),
-            ).reservation = named_resource_spec["Value"]
-        if discrete_resource_spec := res.get("DiscreteResourceSpec"):
-            service_resources.setdefault(
-                discrete_resource_spec["Kind"],
-                ResourceValue(limit=0, reservation=discrete_resource_spec["Value"]),
-            ).reservation = discrete_resource_spec["Value"]
-
-
-def _from_service_settings(
-    settings: list[SimcoreServiceSettingLabelEntry],
-    service_key: ServiceKey,
-    service_version: ServiceVersion,
-    default_service_resources: ServiceResources,
-) -> ServiceResources:
-    # filter resource entries
-    resource_entries = filter(lambda entry: entry.name.lower() == "resources", settings)
-    # get the service resources
-    service_resources = default_service_resources.copy(deep=True)
-    for entry in resource_entries:
-        if not isinstance(entry.value, dict):
-            logger.warning(
-                "resource %s for %s got invalid type, skipping it",
-                f"{entry.dict()!r}",
-                f"{service_key}:{service_version}",
-            )
-            continue
-
-        if nano_cpu_limit := entry.value.get("Limits", {}).get("NanoCPUs"):
-            service_resources["CPU"].limit = nano_cpu_limit / 1.0e09
-        if nano_cpu_reservation := entry.value.get("Reservations", {}).get("NanoCPUs"):
-            service_resources["CPU"].reservation = nano_cpu_reservation / 1.0e09
-        if ram_limit := entry.value.get("Limits", {}).get("MemoryBytes"):
-            service_resources["RAM"].limit = ram_limit
-        if ram_reservation := entry.value.get("Reservations", {}).get("MemoryBytes"):
-            service_resources["RAM"].reservation = ram_reservation
-
-        _parse_generic_resource(
-            entry.value.get("Reservations", {}).get("GenericResources", []),
-            service_resources,
-        )
-
-    return service_resources
+SIMCORE_SERVICE_SETTINGS_LABELS: Final[str] = "simcore.service.settings"
+SIMCORE_SERVICE_COMPOSE_SPEC_LABEL: Final[str] = "simcore.service.compose-spec"
 
 
 @router.get(
     "/{service_key:path}/{service_version}/resources",
-    response_model=dict[ResourceName, ResourceValue],
+    response_model=ServiceResourcesDict,
     **RESPONSE_MODEL_POLICY,
 )
 @cached(
     ttl=DIRECTOR_CACHING_TTL,
     key_builder=lambda f, *args, **kwargs: f"{f.__name__}_{kwargs['service_key']}_{kwargs['service_version']}",
 )
 async def get_service_resources(
```

Comment:
Is sonarcloud not complaining about the cognitivity of this function? I actually separated it in several sub functions to this end. And I see you reverted mostly everything.

Reply:
I've done all this before merging and I kept it. Will have a look at sonarcloud.

Reply:
Now that I've bothered to read what sonarcloud was going on about. I get the point. will refactor it.